packages/venia-ui/lib/util/apolloCache.js

```diff
@@ -0,0 +1,187 @@
+    RevokeCustomerTokenOutput: 'RevokeCustomerTokenOutput',
+    SearchResultPageInfo: 'SearchResultPageInfo',
+    SelectedConfigurableOption: 'SelectedConfigurableOption',
+    SelectedCustomizableOption: 'SelectedCustomizableOption',
+    SelectedCustomizableOptionValue: 'SelectedCustomizableOptionValue',
+    SelectedPaymentMethod: 'SelectedPaymentMethod',
+    SelectedShippingMethod: 'SelectedShippingMethod',
+    SendEmailToFriendOutput: 'SendEmailToFriendOutput',
+    SendEmailToFriendRecipient: 'SendEmailToFriendRecipient',
+    SendEmailToFriendSender: 'SendEmailToFriendSender',
+    SetBillingAddressOnCartOutput: 'SetBillingAddressOnCartOutput',
+    SetGuestEmailOnCartOutput: 'SetGuestEmailOnCartOutput',
+    SetPaymentMethodOnCartOutput: 'SetPaymentMethodOnCartOutput',
+    SetShippingAddressesOnCartOutput: 'SetShippingAddressesOnCartOutput',
+    SetShippingMethodsOnCartOutput: 'SetShippingMethodsOnCartOutput',
+    ShipBundleItemsEnum: 'ShipBundleItemsEnum',
+    ShippingCartAddress: 'ShippingCartAddress',
+    SimpleCartItem: 'SimpleCartItem',
+    SimpleProduct: 'SimpleProduct',
+    SortEnum: 'SortEnum',
+    SortField: 'SortField',
+    SortFields: 'SortFields',
+    StoreConfig: 'StoreConfig',
+    SwatchData: 'SwatchData',
+    SwatchLayerFilterItem: 'SwatchLayerFilterItem',
+    SwatchLayerFilterItemInterface: 'SwatchLayerFilterItemInterface',
+    UpdateCartItemsOutput: 'UpdateCartItemsOutput',
+    UrlRewrite: 'UrlRewrite',
+    UrlRewriteEntityTypeEnum: 'UrlRewriteEntityTypeEnum',
+    VirtualCartItem: 'VirtualCartItem',
+    VirtualProduct: 'VirtualProduct',
+    Website: 'Website',
+    WishlistItem: 'WishlistItem',
+    WishlistOutput: 'WishlistOutput'
+};
+
+/**
+ * The default way the Apollo InMemoryCache stores objects is by using a key
+ * that is a concatenation of the `__typename` and `id` (or `_id`) fields.
+ * For example, "ConfigurableProduct:1098".
+ *
+ * Unfortunately, not all Magento 2 GraphQL Types have an `id` (or `_id`) field.
+ * This function "normalizes" those Type objects by generating a custom unique key for them
+ * that will be used by the Apollo cache.
+ *
+ * @see https://www.apollographql.com/docs/resources/graphql-glossary/#normalization.
+ *
+ * @param {object} A GraphQL Type object.
+ */
+export const cacheKeyFromType = object => {
+    switch (object.__typename) {
+        // Store all implementations of ProductInterface with the same prefix,
+        // and because we can't filter / query by id, use their url_key.
+        case MagentoGraphQLTypes.BundleProduct:
+        case MagentoGraphQLTypes.ConfigurableProduct:
+        case MagentoGraphQLTypes.DownloadableProduct:
+        case MagentoGraphQLTypes.GiftCardProduct:
+        case MagentoGraphQLTypes.GroupedProduct:
+        case MagentoGraphQLTypes.SimpleProduct:
+        case MagentoGraphQLTypes.VirtualProduct:
+            return `${MagentoGraphQLTypes.ProductInterface}:${object.url_key}`;
```

Comment:
Good question!

Remember that the PDP only has a `url_key` to use to find its product. Without grouping these, entries in the cache would look like:

`SimpleProduct:augusta-earrings`
`ConfigurableProduct:claudia-crochet-dress`

Even though the PDP knows `claudia-crochet-dress`, it doesn't know the product's `type` and would therefore be unable to generate the correct key for the cache.

Grouping them all to:

`ProductInterface:augusta-earrings`
`ProductInterface:claudia-crochet-dress`

makes it so that the PDP doesn't have to know the product type.

It works only because the `url_key`s are unique even across product types. If we had simple products and configurable products with the same `url_key`, they would overwrite each other and this wouldn't work.

Reply:
A question was raised this evening and I think it warrants a look: https://magento.slack.com/archives/C7LRPD5FT/p1572299867001800

Basically this caching might not work for multi-store/locale clients since product names are localized.

Reply:
We can't count on any aspect of the product data except the SKU being the same when you switch stores or locales, believe it or not. So breaking the cache when the url_key changes is actually what we want.